core/corehttp/gateway_handler.go

```diff
@@ -386,6 +386,11 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 		webError(w, "ipfs resolve -r "+debugStr(contentPath.String()), err, http.StatusServiceUnavailable)
 		return
 	default:
+		// cache 404s where we know the path does not exist
+		if _, ok := err.(resolver.ErrNoLink); ok {
```

Comment:
```suggestion
		if errors.Is(err, resolver.ErrNoLink{}) {
```
Warning untested code.